Release: 0.9.0

Anton-Latukha opened the pull request:
### 0.9.0

* Changelog started, changes from previous release `0.8.0`:

* Major breaking:
  * Removed instances due to migration to `haskeline >= 0.8 && < 0.9`:
    * `instance MonadException m => MonadException(StateT(HashMap FilePath NExprLoc) m)`
    * `instance MonadException m => MonadException(Fix1T StandardTF m)`

* Minor:
  * Development package neetly exposes majority of `Nixpkgs haskell.lib` features
  * Fixed SDist build
  * Changes in some REPL funtion types, including `main` function
  * Freed from `cryptohash-{md5,sha1,sha256,sha512}` dependencies in favour of `hashing` for the time being
  * New CI:
    * Provides `Nixpkgs haskell.lib` features
    * SDist build
    * Strict build (compilation with `-Werror`) for `GHC < 8.8`
  * Added support for GHC 8.4.4

Anton-Latukha committed:
add CHANGELOG.md

A  CHANGELOG.md

Anton-Latukha committed:
hnix.cabal: version 0.9.0

M  hnix.cabal

sjakobi commented on CHANGELOG.md:
```diff
@@ -0,0 +1,27 @@
+    * `instance MonadException m => MonadException(Fix1T StandardTF m)`
+
+* Minor:
+  * Development package neetly exposes majority of `Nixpkgs haskell.lib` features
```
```suggestion
  * Development package neatly exposes majority of `Nixpkgs haskell.lib` features
```

Is this interesting for the users though? Nix users possibley? (Honest question)

Anton-Latukha replied:
Changelogs are mostly for IT specialists.

Our IT users are Haskell or Nix specialists ... profit.

You can basically open a `default.nix`, change keys how do you like it, and use simple `nix-build` to check that the project builds. I am interested is it affects `ghcid` or `HIE`, IDK at the moment.

Anton-Latukha replied:
I remembered now, I already know that `default.nix` affects the `HIE` directly, at least in `all-HIEs` `NixOS` setup that I documented.

Anton-Latukha replied:
With `build.sh` by its idea and the code. It should be possible to just export shell variables into `env` and run `build.sh` anywhere, and it would import and build with the same options.

sjakobi commented on CHANGELOG.md:
```diff
@@ -0,0 +1,27 @@
+* Minor:
+  * Development package neetly exposes majority of `Nixpkgs haskell.lib` features
+  * Fixed SDist build
+  * Changes in some REPL funtion types, including `main` function
```
I don't believe that users will care about this –  the repl module is not exposed.

```suggestion
```

Anton-Latukha replied:
That is why it is in minor.

sjakobi replied:
Yeah. It's not even a minor user-facing change IMHO.

sjakobi commented on CHANGELOG.md:
```diff
@@ -0,0 +1,27 @@
+  * Development package neetly exposes majority of `Nixpkgs haskell.lib` features
+  * Fixed SDist build
+  * Changes in some REPL funtion types, including `main` function
+  * Freed from `cryptohash-{md5,sha1,sha256,sha512}` dependencies in favour of `hashing` for the time being
```
We already depended on `hashing` before.

```suggestion
  * Removed dependencies on `cryptohash-{md5,sha1,sha256,sha512}`
```

Anton-Latukha replied:
Ok.

Anton-Latukha replied:
I did not stated that `hashing` is a new dependency, so https://github.com/haskell-nix/hnix/pull/627#discussion_r439861918 is a straw man argument.
_"Freed from `cryptohash-{md5,sha1,sha256,sha512}` dependencies in favour of `hashing` for the time being."_ does not mean that `hashing` is a newly added dependency.

Anton-Latukha replied:
As you said previously in PR of it - the `hashing` is slow and only `hnix` uses it as a dependency. So I explained in changelog that it is "for the time being", so people would not run to us explaining that `hashing` is slow and only `hnix` uses it as a dependency.

sjakobi replied:
> I did not stated that `hashing` is a new dependency

I thought the sentence sounds a bit like that. The use of `hashing` seems just irrelevant – it's not a change.

Anton-Latukha replied:
Well, it sounds like that, because we need to use it instead of the... well, Ok.

Thank you.

sjakobi commented on CHANGELOG.md:
```diff
@@ -0,0 +1,27 @@
+  * New CI:
+    * Provides `Nixpkgs haskell.lib` features
+    * SDist build
+    * Strict build (compilation with `-Werror`) for `GHC < 8.8`
```
This was amazing work, but not a user-facing change…

Anton-Latukha replied:
It is just project state, quality news, news for devs, maybe then we can just remove `Minor`, since dependencies on `cryptohash-{md5,sha1,sha256,sha512}` are also not user-facing changes.

Anton-Latukha replied:
I wrote it more explicitly to get a feeling of what can be slashed on next recurses.

I feel like most of `minor` changes in the project can be skipped.

Only `major` and `medium` of extensions in API functionality should be mentioned.

sjakobi replied:
> dependencies on `cryptohash-{md5,sha1,sha256,sha512}` are also not user-facing changes.

I think that's somewhat user-facing, since their dependency trees change.

Anton-Latukha replied:
Then we may argue that everything is user-facing, since user uses the project code and dependencies of the project inside an OS on some machine (obviously _reductio ad absurdum_ here).

Did you ever cared when unknown to you the Haskell libraries changed lib dependencies or introduced custom prelude, in libraries/modules you never looked in the code of? Or you are awake at night monitoring/controlling all the underlying changes in all Hackage libraries (obviously _reductio ad absurdum_ here).

Effectiveness - is to not care and trust Haskell community and ecosystem, it is one of the best ecosystems to trust and draw your effectiveness from trust and not caring about details.

sjakobi commented on CHANGELOG.md:
```diff
@@ -0,0 +1,27 @@
+    * Provides `Nixpkgs haskell.lib` features
+    * SDist build
+    * Strict build (compilation with `-Werror`) for `GHC < 8.8`
+  * Added support for GHC 8.4.4
```
0.8.0 was compatible with GHC 8.4.4 too, AFAIK.

```suggestion
```

Anton-Latukha replied:
Well if to call the state that we did not know or cared if `8.4.4` worked - a support.

For example, maybe we compile on ARM `aarch64`, we do not know thou, so does project claims support of it?

Project can not claim to support something, until project pays attention and effort to be sure to support it with every release, and officially stated the support in the first place.

sjakobi replied:
The main thing that signal GHC support for Haskell users, are the `base` bounds. The specific lower bound `>= 4.11` signals support for GHC 8.4.

Anton-Latukha replied:
If `base` is the main indicator - then we should test that and so then we should claim the support of the `base` versions, not GHCs.

In project description there may be no bounds at all, they can be open (as bounds in widely used `stack`, if it is an indicator - does it mean `stack` infinitely supports infinite future?) or bounds just may be plain forgotten to be updated or wrong or have errors we are unaware of.

What project claim costs, if `base` bounds are set but the project fails to assemble/compile?

Well, for me the signal is successful CI builds, purely functional builds in Nix show really does project really compiles with GHC 8.4.4 or 8.8.3.

sjakobi commented on CHANGELOG.md:
```diff
@@ -0,0 +1,27 @@
+    * `instance MonadException m => MonadException(StateT(HashMap FilePath NExprLoc) m)`
+    * `instance MonadException m => MonadException(Fix1T StandardTF m)`
+
+* Minor:
```
```suggestion
* Minor:
  * Support for GHC 8.10
```

Anton-Latukha replied:
Do we?

sjakobi replied:
The `haskeline` update allowed us to build with GHC 8.10 too. I'd call the [`liftTyped` issue](https://github.com/haskell-nix/hnix/issues/592#issuecomment-634665426) a known bug – it won't affect existing code though. Do you think we should point that out in the changelog?

Anton-Latukha replied:
CI never successfully compiled a project with GHC 8.10, GHC 8.10 support report #592 still open, so we can not claim project support of it, until every PR/release is not checked to comply with the GHC 8.10.

sjakobi replied:
> CI never successfully compiled a project with GHC 8.10

I thought that was just a nixpkgs problem, not a property of the Cabal package that we're publishing here.

Anton-Latukha replied:
Honesly, since other people who already know the theme are on that issue, I did not looked into it.

In reality, Nixpkgs current default compiler is `GHC 8.8.3`, so all `haskellPackages` infrastructure is suited for it, no wonder it does not support `GHC 8.10`. It was my main case, but I did not looked into it further, and I still didn't looked and understood/found the cause, so reserved judgment.

The proper way to solve this - is to have a separate build that does how project states in readme. `nix-shell -> cabal v2-build` - that would be a quite other build (with different Cabal parameters) from what CI did&does (`nix-build`).
And build without Nix only with Cabal - is the other build.

And then the question why we have all those Nix files and being an Nix interpreter, but those arguments does not hold to the core issue that this is in truth the Haskell project first, and it is Nix interpreter is the second here.

Anton-Latukha replied:
We should have some basic Cabal build, that way we would see where we hit the Nixpkgs internal issues. (rhetorical question) - You feel here should be multicheck that reports that something in fact works, but hits a Nixpkgs code somewhere?

Anton-Latukha replied:
Again, only when GHC 8.10 checks in CI, I can claim its support, because only then I would be sure and able to merge the code only when that code supports GHC 8.10, and be sure that all others write and merge GHC 8.10 compatible code.

If you want to claim the support - we can add it then.

sjakobi reviewed (commented):
Many thanks @Anton-Latukha! :)

I have a few comments on the changelog. The general question I believe is what's the criterion for inclusion in the changelog. When I read changelogs as a user, I mainly want to know about the changes that might be relevant for my use of the library.

Anton-Latukha committed:
hnix.cabal: extra-source-files: add CHANGELOG.md

M  hnix.cabal

Anton-Latukha committed:
hnix.cabal: set {lens-family{,-core},template-haskell} bounds

Set current upper bound of lenses

`template-haskell` bounds derived from the git commit date and current bound,
then lowered low bound to include GHC 8.4

M  hnix.cabal

M  hnix.cabal
